trulens_eval/trulens_eval/feedback.py

```diff
@@ -1,5 +1,353 @@
+feedback functions likewise need to belong to a package that can be imported.
+Critically, classes defined locally in a notebook will not be importable this
+way.
+
+## Specifying Arguments
+
+The mapping between app/records to feedback implementation arguments is
+specified by the `on...` methods of the `Feedback` objects. The general form is:
+
+```python
+feedback: Feedback = feedback.on(argname1=selector1, argname2=selector2, ...)
+```
+
+That is, `Feedback.on(...)` returns a new `Feedback` object with additional
+argument mappings, the source of `argname1` is `selector1` and so on for further
+argument names. The types of `selector1` is `JSONPath` which we elaborate on in
+the further sections.
+
+If argument names are ommitted, they are taken from the feedback function
+implementation signature in order. That is, 
+
+```python
+...on(argname1=selector1, argname2=selector2)
+```
+
+and
+
+```python
+...on(selector1, selector2)
+```
+
+are equivalent assuming the feedback implementation has two arguments,
+`argname1` and `argname2`, in that order.
+
+Several utility methods starting with `.on` provide shorthands:
+
+- `on_input(arg) == on_prompt(arg: Optional[str])` -- both specify that the next
+    unspecified argument or `arg` should be the main app input.
+
+- `on_output(arg) == on_response(arg: Optional[str])` -- specify that the next
+    argument or `arg` should be the main app output.
+
+- `on_input_output() == on_input().on_output()` -- specifies that the first
+    two arguments of implementation should be the main app input and main app
+    output, respectively.
+
+- `on_default()` -- depending on signature of implementation uses either
+    `on_output()` if it has a single argument, or `on_input_output` if it has
+    two arguments.
+
+Some wrappers include additional shorthands:
+
+### llama_index-specific selectors
+
+- `TruLlama.select_source_nodes()` -- outputs the selector of the source
+    documents part of the engine output.
+
+### Running Feedback
+
+Feedback implementations are simple callables that can be run on any arguments
+matching their signatures. However, once wrapped with `Feedback`, they are meant
+to be run on outputs of app evaluation (the "Records"). Specifically,
+`Feedback.run` has this definition:
+
+```python
+def run(self, 
+    app: Union[AppDefinition, JSON], 
+    record: Record
+) -> FeedbackResult:
+```
+
+That is, the context of a Feedback evaluation is an app (either as
+`AppDefinition` or a JSON-like object) and a `Record` of the execution of the
+aforementioned app. Both objects are indexable using "Selectors". By indexable
+here we mean that their internal components can be specified by a Selector and
+subsequently that internal component can be extracted using that selector.
+Selectors for Feedback start by specifying whether they are indexing into an App
+or a Record via the `__app__` and `__record__` special
+attributes (see **Selectors** section below).
```

Comment:
probably just remove this, i don't think this detail is something the user is going to need. Plus we've mentioned the selectors a few times already

Reply:
or; make it super concise. like

Both Applications and Records are serialized into JsonPath and can be selected upon.

Reply:
I think the part about selecting from record or app needs to be made explicit and this is where it is done.